aptos-move/aptos-vm/src/move_vm_ext/respawned_session.rs

```diff
@@ -0,0 +1,127 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    aptos_vm_impl::AptosVMImpl,
+    data_cache::{AsMoveResolver, StorageAdapter},
+    move_vm_ext::{SessionExt, SessionId},
+};
+use anyhow::{bail, Result};
+use aptos_aggregator::transaction::ChangeSetExt;
+use aptos_gas::ChangeSetConfigs;
+use aptos_state_view::{StateView, StateViewId, TStateView};
+use aptos_types::{
+    state_store::{
+        state_key::StateKey, state_storage_usage::StateStorageUsage, state_value::StateValue,
+    },
+    write_set::{TransactionWrite, WriteSet},
+};
+use move_core_types::vm_status::{err_msg, StatusCode, VMStatus};
+
+/// We finish the session after the user transaction is done running to get the change set and
+/// charge gas and storage fee based on it before running storage refunds and the transaction
+/// epilogue. The latter needs to see the state view as if the change set is applied on top of
+/// the base state view, and this struct implements that.
+#[ouroboros::self_referencing]
+pub struct RespawnedSession<'r, 'l> {
+    state_view: ChangeSetStateView<'r>,
+    #[borrows(state_view)]
+    #[covariant]
+    resolver: StorageAdapter<'this, ChangeSetStateView<'r>>,
+    #[borrows(resolver)]
+    #[not_covariant]
+    session: Option<SessionExt<'this, 'l>>,
```

Comment:
why is this option?

Reply:
I need to `take` and consume it on line #62, to get the new change set before destructing `self`.